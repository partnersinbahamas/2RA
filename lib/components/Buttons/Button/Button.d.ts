import { default as React } from 'react';
import { default as TPropsButton } from '../../utils/types/props/button';
import { TSize, TStile } from '../../utils/types/types';

export type TProps = TPropsButton & {
    size?: TSize;
    type?: TButtonType;
    stile?: TStile;
    style?: React.CSSProperties;
    disabled?: boolean;
    /**
     * TError
     * @type string | boolean
     */
    error?: TError;
    onClick?: () => void;
    className?: any;
    title?: string;
    body?: any;
    disablePadding?: boolean;
    nonTitled?: boolean;
};
export declare const Button: React.FC<TProps>;
