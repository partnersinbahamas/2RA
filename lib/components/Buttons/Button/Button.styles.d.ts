import { TProps as TButtonProps } from './Button';

declare const ButtonComponent: import('styled-components').IStyledComponent<"web", import('styled-components/dist/types').Substitute<import('react').DetailedHTMLProps<import('react').ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement>, TButtonProps>>;
export { ButtonComponent };
