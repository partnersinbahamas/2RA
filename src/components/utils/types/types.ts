export type TSize = 'small' | 'medium' | 'large';

export type TStile = 'primary' | 'default' | 'mute';

export type TStatus = "online" | "default" | "busy" | "not-here" | "sleep" | "primary";
export type TStringOrStatus = TStatus | string;
