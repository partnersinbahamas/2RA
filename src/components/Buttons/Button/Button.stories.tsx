import { Meta, StoryObj } from '@storybook/react';
import { Button } from './Button';
import { story } from '../../utils/variables/story';

const meta = {
  title: 'Athomic/Buttons/Button',
  tags: ['autodocs'],
  component: Button,
  argTypes: {
    size: story.size,
    type: story.buttonType,
    onClick: story.onClick,
    className: story.className,
    title: story.title,
    disabled: story.disabled,
  },
  parameters: {
    layout: 'centered',
    docs: {
      description: {
        component: `
Please to style your custom input component use the mockup bellow.
### Guide styles
  .scss

  .module.scss

  use the "button" prefix to make the class more specific

    button.button {
      &[disabled] {};
    }
    `,
      },
    },
  },
} satisfies Meta<typeof Button>;
export default meta;

type TStory = StoryObj<typeof meta>;

export const Default: TStory = {};
export const DefaultDisabled: TStory = { args: { disabled: true } };

export const Primary: TStory = { args: { stile: 'primary' } };
export const PrimaryDisabled: TStory = {
  args: {
    stile: 'primary',
    disabled: true,
  },
};

export const Mute: TStory = { args: { stile: 'mute' } };
export const MuteDisabled: TStory = {
  args: {
    stile: 'mute',
    disabled: true,
  },
};
