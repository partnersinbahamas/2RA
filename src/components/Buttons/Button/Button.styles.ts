import styled from 'styled-components';
import { TProps as TButtonProps } from './Button';
import {
  BLUE_200,
  BLUE_400,
  BLUE_500,
  COOL_GRAY_200,
  COOL_GRAY_300,
  COOL_GRAY_400,
  COOL_GRAY_500,
  CYAN_400,
  WHITE,
} from '../../Colors/colors';
import usePrime from '../../../hooks/usePrime';
import { TStile } from '../../utils/types/types';
import defaultProps from '../../utils/variables/defaultProps';
import useMute from '../../../hooks/useMute';
import { fontTuraRegularBold } from '../../utils/variables/fonts';

const prime = (stile: TStile | undefined) =>
  usePrime(stile || defaultProps.stile).primeState;
const mute = (stile: TStile | undefined) =>
  useMute(stile || defaultProps.stile).muteState;

// couldn't use prime hook cause usefined & defaultProps;
const isDefault = (stile: TStile | undefined) =>
  !mute(stile) && stile !== 'primary';

const ButtonComponent = styled.button<TButtonProps>`
  ${({ stile }) =>
    !mute(stile) &&
    `
    ${fontTuraRegularBold}
      border-radius: 3em;
      cursor: pointer;
      display: inline-block;
      line-height: 1;

      &[disabled] {
        cursor: not-allowed;
      };
    `}

  ${({ stile }) =>
    isDefault(stile) &&
    `
      background-color: ${COOL_GRAY_200};
      border: 2px solid ${COOL_GRAY_300};
      color: ${COOL_GRAY_400};

      &[disabled] {
        background-color: ${COOL_GRAY_400};
        border-color: ${COOL_GRAY_500};
        color: ${COOL_GRAY_200};
      };
    `}

  ${({ stile }) =>
    prime(stile) &&
    `
    border: 2px solid ${BLUE_400};
    background-color: ${CYAN_400};
    color: ${WHITE};

    &[disabled] {
      background-color: ${BLUE_400};
      border-color: ${BLUE_500};
      color: ${BLUE_200};
    };
  `}
`;

export { ButtonComponent };
