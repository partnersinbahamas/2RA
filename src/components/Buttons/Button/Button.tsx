import React from 'react';
import classNames from 'classnames';

import defaultProps from '../../utils/variables/defaultProps';
import { TSize, TStile } from '../../utils/types/types';
import { ButtonComponent } from './Button.styles';
import styles from './Button.module.scss';

export type TProps = React.ComponentProps<'button'> & {
  size?: TSize;
  type?: TButtonType;
  stile?: TStile;
  style?: React.CSSProperties;
  disabled?: boolean;
  onClick?: () => void;
  className?: any;
  title?: string;
  body?: any;
  disablePadding?: boolean;
  nonTitled?: boolean;
};

export const Button: React.FC<TProps> = ({
  size = defaultProps.size,
  type = defaultProps.buttonType,
  stile = defaultProps.stile,
  disabled,
  className,
  title = 'Button',
  body,
  disablePadding,
  nonTitled,
  ...props
}) => {
  return (
    <ButtonComponent
      {...props}
      type={type}
      disabled={disabled}
      className={classNames(styles[size], className)}
      stile={stile}
      style={{
        ...props.style,
        ...(disablePadding ? { padding: '0' } : {}),
      }}
    >
      {!nonTitled && title}
      {body && !!Object.keys(body).length && body}
    </ButtonComponent>
  );
};
