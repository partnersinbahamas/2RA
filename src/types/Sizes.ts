export type TSizes = 'small' | 'medium' | 'large';