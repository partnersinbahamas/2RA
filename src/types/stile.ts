export type TStile = 'primary' | 'default' | 'mute';