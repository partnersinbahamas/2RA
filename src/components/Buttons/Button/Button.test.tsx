import { TStile } from "../../utils/types/types";
import { Button } from "./Button";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe('Button', () => {
  const user = userEvent.setup();

  const props = {
    title: 'Button',
    onClick: jest.fn(),
    body: <div data-testid="body-element" />
  };

  it('renders correctly', () => {
    const { container } = render(<Button />);
    const buttonElement = screen.getByRole('button');
    expect(buttonElement).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('renders with title', () => {
    const { container } = render(<Button {...props} />);

    const buttonElement = screen.getByRole('button', { name: props.title });
    expect(buttonElement).toHaveTextContent(props.title as string);
    expect(container).toMatchSnapshot();
  });

  it.each(['default', 'primary', 'mute'])('renders with stile', (stile) => {
    const { container } = render(<Button {...props} stile={stile as TStile} />);

    const buttonElement = screen.getByRole('button', { name: props.title });
    expect(buttonElement).toHaveAttribute('stile', stile);
    expect(container).toMatchSnapshot();
  });

  it('should render click handler', async () => {
    const { container } = render(<Button {...props}/>);

    const buttonElement = screen.getByRole('button', { name: props.title });
    await user.click(buttonElement);

    expect(props.onClick).toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  it('should renders with body', async () => {
    const { container } = render(<Button {...props} />);

    const bodyElement = screen.queryByTestId('body-element');
    expect(bodyElement).not.toBeNull();
    expect(bodyElement).toBeVisible();

    expect(container).toMatchSnapshot();
  });
})