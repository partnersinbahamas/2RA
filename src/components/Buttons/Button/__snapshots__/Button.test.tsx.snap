// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<div>
  <button
    class="sc-beySPh crhrOy medium"
    stile="default"
    type="button"
  >
    Button
  </button>
</div>
`;

exports[`Button renders with stile 1`] = `
<div>
  <button
    class="sc-beySPh mHYUy medium"
    stile="mute"
    type="button"
  >
    Button
    <div
      data-testid="body-element"
    />
  </button>
</div>
`;

exports[`Button renders with stile 2`] = `
<div>
  <button
    class="sc-beySPh crhrOy medium"
    stile="default"
    type="button"
  >
    Button
    <div
      data-testid="body-element"
    />
  </button>
</div>
`;

exports[`Button renders with stile 3`] = `
<div>
  <button
    class="sc-beySPh hoqXzN medium"
    stile="primary"
    type="button"
  >
    Button
    <div
      data-testid="body-element"
    />
  </button>
</div>
`;

exports[`Button renders with title 1`] = `
<div>
  <button
    class="sc-beySPh crhrOy medium"
    stile="default"
    type="button"
  >
    Button
    <div
      data-testid="body-element"
    />
  </button>
</div>
`;

exports[`Button should render click handler 1`] = `
<div>
  <button
    class="sc-beySPh crhrOy medium"
    stile="default"
    type="button"
  >
    Button
    <div
      data-testid="body-element"
    />
  </button>
</div>
`;

exports[`Button should renders with body 1`] = `
<div>
  <button
    class="sc-beySPh crhrOy medium"
    stile="default"
    type="button"
  >
    Button
    <div
      data-testid="body-element"
    />
  </button>
</div>
`;
